File: code/ch07/ruby/humble_view.rb
# §7.3.3 Keep Views Simple — fat view vs. humble view over a view-model
# (clinic scheduler: invoice list with an overdue badge).
require "date"

Invoice = Struct.new(:patient_name, :sent_on, :paid)

# --- Before: fat view ---------------------------------------------------------
class InvoiceWidget
  def render(invoice, today)
    text = invoice.patient_name
    if !invoice.paid && (today - invoice.sent_on) > 30
      text += " — OVERDUE"             # a business rule, trapped on screen
    end
    text
  end
end

ana = Invoice.new("Ana", Date.new(2026, 6, 1), false)
today = Date.new(2026, 7, 2)
fat = InvoiceWidget.new.render(ana, today)
raise unless fat == "Ana — OVERDUE"

# --- After: humble view -------------------------------------------------------
def overdue?(invoice, today)             # the rule, extracted where tests reach it
  !invoice.paid && (today - invoice.sent_on) > 30
end

def invoice_view_model(invoice, today)
  badge = overdue?(invoice, today) ? " — OVERDUE" : ""
  { text: invoice.patient_name + badge }
end

class HumbleInvoiceWidget                # the book shows this as InvoiceWidget,
  def render(vm)                         # rewritten; renamed so both versions
    vm[:text]                            # run in one file
  end
end

def test_unpaid_31_days_is_overdue
  ana = Invoice.new("Ana", Date.new(2026, 6, 1), false)
  raise unless overdue?(ana, Date.new(2026, 7, 2))
end

raise unless HumbleInvoiceWidget.new.render(invoice_view_model(ana, today)) == fat
raise if overdue?(Invoice.new("Ben", Date.new(2026, 6, 15), false), today)
raise if overdue?(Invoice.new("Cho", Date.new(2026, 5, 1), true), today)
test_unpaid_31_days_is_overdue
puts "fat view and humble view render identically; rule tested directly"

File: code/ch07/javascript/humble_view.js
// §7.3.3 Keep Views Simple — fat view vs. humble view over a view-model
// (clinic scheduler: invoice list with an overdue badge).
const assert = require("node:assert");

const MS_PER_DAY = 86_400_000;

// --- Before: fat view --------------------------------------------------------
class InvoiceWidget {
  render(invoice, today) {
    let text = invoice.patientName;
    if (!invoice.paid && (today - invoice.sentOn) / MS_PER_DAY > 30) {
      text += " — OVERDUE";              // a business rule, trapped on screen
    }
    return text;
  }
}

const ana = { patientName: "Ana", sentOn: new Date("2026-06-01"), paid: false };
const today = new Date("2026-07-02");
const fat = new InvoiceWidget().render(ana, today);
assert.strictEqual(fat, "Ana — OVERDUE");

// --- After: humble view ------------------------------------------------------
function isOverdue(invoice, today) {     // the rule, extracted where tests reach it
  return !invoice.paid && (today - invoice.sentOn) / MS_PER_DAY > 30;
}

function invoiceViewModel(invoice, today) {
  const badge = isOverdue(invoice, today) ? " — OVERDUE" : "";
  return { text: invoice.patientName + badge };
}

class HumbleInvoiceWidget {              // the book shows this as InvoiceWidget,
  render(vm) {                           // rewritten; renamed so both versions
    return vm.text;                      // run in one file
  }
}

function testUnpaid31DaysIsOverdue() {
  const ana = { patientName: "Ana", sentOn: new Date("2026-06-01"), paid: false };
  assert(isOverdue(ana, new Date("2026-07-02")));
}

assert.strictEqual(new HumbleInvoiceWidget().render(invoiceViewModel(ana, today)), fat);
const ben = { patientName: "Ben", sentOn: new Date("2026-06-15"), paid: false };
const cho = { patientName: "Cho", sentOn: new Date("2026-05-01"), paid: true };
assert(!isOverdue(ben, today) && !isOverdue(cho, today));
testUnpaid31DaysIsOverdue();
console.log("fat view and humble view render identically; rule tested directly");
